spirv_msl.cpp

```diff
@@ -9626,6 +9656,75 @@ void CompilerMSL::emit_instruction(const Instruction &instruction)
 		break;
 	}
 
+	case OpSDot:
+	case OpUDot:
+	case OpSUDot:
+	{
+		uint32_t result_type = ops[0];
+		uint32_t id = ops[1];
+		uint32_t vec1 = ops[2];
+		uint32_t vec2 = ops[3];
+
+		auto& input_type1 = expression_type(vec1);
+		auto& input_type2 = expression_type(vec2);
+
+		string vec1input, vec2input;
+		if ((input_type1.basetype == SPIRType::BaseType::UInt || input_type1.basetype == SPIRType::BaseType::Int) && input_type1.vecsize == 1)
+		{
+			string type = opcode == OpSDot || opcode == OpSUDot ? "char4" : "uchar4";
+			vec1input = join("as_type<", type, ">(", to_expression(vec1), ")");
+		} else
+			vec1input = to_expression(vec1);
+
+		if ((input_type2.basetype == SPIRType::BaseType::UInt || input_type2.basetype == SPIRType::BaseType::Int) && input_type2.vecsize == 1)
+		{
+			string type = opcode == OpSDot ? "char4" : "uchar4";
+			vec2input = join("as_type<", type, ">(", to_expression(vec2), ")");
+		} else
+			vec2input = to_expression(vec2);
+
+		auto& type = get<SPIRType>(result_type);
+		auto input_width = input_type1.vecsize == 1 ? 4 : input_type1.vecsize;
+		string exp = join("reduce_add(", type_to_glsl(type), std::to_string(input_width), "(", vec1input, ") * ", type_to_glsl(type), std::to_string(input_width), "(", vec2input, "))");
```

Comment:
std::to_string(input_width) is not needed in join(). Just use the integer directly.